### PR TITLE
perf: make CLI streaming render incremental instead of O(n^2)

### PR DESCRIPTION
**Pull request by AgarwalManas**
## Summary

Addresses the Rust CLI portion of #10075 (the Electron and Ink frontends
have separate root causes in separate codebases and are left for separate
PRs).

Streamed output re-did work proportional to the whole accumulated message
on every token delta, in two places:

- `print_markdown_raw` constructed a fresh `bat::PrettyPrinter` per flush,
  re-deserializing bat's bundled syntax/theme assets each call, roughly
  once per token during prose — serialized in the same loop that pulls the
  SSE stream. The printer is now cached per thread; `print()` drains queued
  inputs but leaves the loaded assets reusable.

- `MarkdownBuffer::find_safe_end` re-scanned the entire un-drained buffer
  on every push. Inside a long fenced code block the buffer cannot drain
  until the closing fence, so each delta re-scanned the whole accumulated
  block — O(n²) in block length. The scan now carries a checkpoint across
  non-draining pushes so each delta scans only the new tail. A drain resets
  the checkpoint (the previous algorithm reinterprets the remainder from
  scratch, and that behavior is preserved exactly), and checkpoints are
  never persisted into the trailing region of whitespace/fence-only lines,
  whose interpretation can still change as the buffer grows.

### Testing

- All pre-existing drain-boundary tests pass unchanged.
- New: a differential fuzz test enforcing byte-identical behavior between
  the checkpointed scan and a full rescan (a zeroed checkpoint degenerates
  to the previous algorithm) across randomized streams and chunk
  boundaries — 500 cases in the suite; a 5,000-case sweep also passed.
  This test caught two checkpoint bugs during development (fence runs
  growing at the buffer edge; drains changing how the remainder parses);
  both are fixed and covered.
- New checkpoint edge-case tests: fence markers split across pushes,
  blank-line lookahead, chunking-reassembly property.
- Benchmark (4,000-line fenced block streamed in 8-byte chunks, release
  build, N=10 runs each): **1.19 s mean before → 2.4 ms mean after**. An
  `#[ignore]`d test (`large_code_block_streams_in_linear_time`) reproduces
  this: `cargo test -p goose-cli --release -- --ignored large_code_block --nocapture`
- Not tested: an end-to-end session against a live high-token-rate
  provider.

### Related Issues

Relates to #10075

### Screenshots/Demos (for UX changes)

N/A — rendering performance only; output bytes are unchanged (enforced by
the differential test).

### CODE REVIEW

File: crates/goose-cli/src/session/streaming_buffer.rs [hunk 1]
```diff
@@ -298,20 +315,60 @@ impl MarkdownBuffer {
...
+            None => 0,
+        };
+        while limit > 0 {
+            let prev_start = match self.buffer[..limit - 1].rfind('\n') {
+                Some(nl) => nl + 1,
+                None => 0,
+            };
+            let reinterpretable = bytes[prev_start..limit]
+                .iter()
+                .all(|&c| matches!(c, b'`' | b'~' | b' ' | b'\t' | b'\r' | b'\n'));
```
[the remark this answers is not in the document]
Comment (AgarwalManas) [reply crates/goose-cli/src/session/streaming_buffer.rs:341]: @codex Good catch on the corner case, but measured it and this PR doesn't regress it - it's identical to main there. Streaming a fenced block of N blank lines line-by-line (release build): before this PR 134 ms / 1.00 s / 8.0 s for N = 1000/2000/4000; with this PR 136 ms / 1.02 s / 7.9 s. The dominant cost in that input is the parser's per-blank-line trim_start lookahead in check_code_fence, which predates this change.

Checkpoints genuinely can't advance into a whitespace/fence-only tail without changing behavior: under the current parse rules a blank line's interpretation depends on the still-growing tail (a later closing fence actually closes the block at the first blank line above it), so persisting a checkpoint inside that run changes drain boundaries. The differential fuzz test in this PR exists precisely because two subtler "advance anyway" variants produced real divergences. Caching the limit computation doesn't help either — the walk-back isn't the dominant cost (see numbers above).

Net: this PR removes the quadratic behavior for content-bearing blocks and leaves the blank-heavy pathological case exactly as it is on main. Making that case fast would mean changing the parser's blank-line/fence semantics, which I'd rather not fold into a perf PR - happy to file it as a follow-up issue if there's interest.

[the remark this answers is not in the document]
Comment (AgarwalManas) [reply crates/goose-cli/src/session/streaming_buffer.rs:341]: @codex address that feedback

[the remark this answers is not in the document]
Comment (AgarwalManas) [reply crates/goose-cli/src/session/streaming_buffer.rs:341]: @codex address that feedback

**Comment by johanndrews** [comment]
Independently reproduced and measured on macOS (Apple Silicon, rustc 1.96.1, `--release`).

Probe: `MarkdownBuffer` fed a fenced code block in 8-byte deltas (the streaming shape from #10075), timing the full push loop.

| block size | before (`2fd41b9a7^`) | after (`2fd41b9a7`) |
|---|---|---|
| 500 lines | 37 ms | 2 ms |
| 1 000 lines | 89 ms | 1 ms |
| 2 000 lines | 343 ms | 2 ms |
| 4 000 lines | 1 418 ms | 3 ms |

Before, each doubling roughly quadruples the time — the O(n²) shape. After, it is flat. Prose (which drains on every delta) stays at 0–1 ms in both, so the fast path is unaffected. Both runs drain the same byte count (1119/1120), and `cargo test -p goose-cli --lib streaming_buffer` passes 36/36 including `incremental_scan_matches_full_rescan_on_random_streams`.

One observation, not a blocker: caching the printer in a thread-local forces the input through `Cursor::new(content.as_bytes().to_vec())` because of the `'static` bound, so each flush copies the content. Against the ~6.8 ms of asset deserialisation it saves, that is noise — noting it for the record.

**Review by filipkujawa — APPROVED** [review]
nice!
